document_merge_service/api/authentication.py

```diff
@@ -61,49 +62,34 @@ def get_jwt_value(self, request):
 
         return auth[1]
 
-    def get_json(self, url):
-        response = requests.get(url, verify=settings.OIDC_VERIFY_SSL)
-        response.raise_for_status()
-        return response.json()
-
-    def decode_token(self, token, key):
-        return jwt.decode(
-            token=token,
-            key=key,
-            options=settings.OIDC_VALIDATE_CLAIMS_OPTIONS,
-            algorithms=[settings.OIDC_VERIFY_ALGORITHM],
-            audience=settings.OIDC_CLIENT,
+    def get_userinfo(self, token):
+        response = requests.get(
+            settings.OIDC_USERINFO_ENDPOINT,
+            verify=settings.OIDC_VERIFY_SSL,
+            headers={"Authorization": f"Bearer {smart_text(token)}"},
         )
 
-    def keys(self):
-        if settings.OIDC_VERIFY_ALGORITHM.startswith("RS"):
-            return self.get_json(settings.OIDC_JWKS_ENDPOINT)
-        return settings.OIDC_SECRET_KEY
+        try:
+            response.raise_for_status()
+        except requests.HTTPError as e:
+            raise exceptions.AuthenticationFailed(
+                f"Retrieving userinfo from {settings.OIDC_USERINFO_ENDPOINT} "
+                f"failed with error '{str(e)}'."
+            )
+
+        return response.json()
 
     def authenticate(self, request):
-        jwt_value = self.get_jwt_value(request)
-        if jwt_value is None:
+        token = self.get_bearer_token(request)
+        if token is None:
             return None
 
-        for retry in range(2):
-            try:
-                decoded_token = self.decode_token(
-                    jwt_value,
-                    cache.get_or_set("authentication.keys", self.keys, timeout=None),
-                )
-            except ExpiredSignatureError:
-                msg = _("Invalid Authorization header. JWT has expired.")
-                raise exceptions.AuthenticationFailed(msg)
-            except JWTError:
-                if retry == 0 and settings.OIDC_VERIFY_ALGORITHM.startswith("RS"):
-                    # try again with refreshed keys
-                    cache.delete("authentication.keys")
-                    continue
-
-                msg = _("Invalid Authorization Token.")
-                raise exceptions.AuthenticationFailed(msg)
-
-        return OIDCUser(decoded_token), decoded_token
+        userinfo_method = functools.partial(self.get_userinfo, token=token)
+        userinfo = cache.get_or_set(
```

Comment:
Couldn't this be done a bit prettier by using a cache decorator?

Reply:
django doesn't have a decorator for this. So either writing one ourselves or use a library such as https://github.com/peterbe/django-cache-memoize

We only use it in one place so I think this is a bit of an overkill but certainly something to consider once caching is used in several places.

Reply:
Fair point